src/shinylive.ts

```diff
@@ -0,0 +1,512 @@
+          return { name, content };
+      }
+    })
+  );
+
+  await createAndOpenShinyliveApp(files, isPythonApp ? "py" : "r");
+}
+
+/**
+ * Read a directory recursively and return a list of all files as `vscode.Uri`
+ * objects. Directories whose names start with `_` or `.` are ignored.
+ *
+ * @async
+ * @param {vscode.Uri} uri A `vscode.Uri` of any type. If not a directory, the
+ * function returns an array with the single `vscode.Uri` object.
+ * @returns {Promise<vscode.Uri[] | void>}
+ */
+async function readDirectoryRecursively(
+  uri: vscode.Uri
+): Promise<vscode.Uri[] | void> {
+  const pathStat = await vscode.workspace.fs.stat(uri);
+  if (pathStat.type !== vscode.FileType.Directory) {
+    return [uri];
+  }
+
+  // Treat `_dir/` or `.dir/` as hidden directories and discard them
+  if (["_", "."].some((chr) => path.basename(uri.path).startsWith(chr))) {
+    return;
+  }
+
+  const files: vscode.Uri[] = [];
+  const filesInDir = await vscode.workspace.fs.readDirectory(uri);
+
+  for (const [name, type] of filesInDir) {
+    if (type === vscode.FileType.Directory) {
+      const subdir = vscode.Uri.file(`${uri.path}/${name}`);
+      const subdirFiles = await readDirectoryRecursively(subdir);
+      if (subdirFiles) {
+        files.push(...subdirFiles);
+      }
+    } else {
+      files.push(vscode.Uri.file(`${uri.path}/${name}`));
+    }
+  }
+
+  return files;
+}
+
+/**
+ * Consult the user's preferred open action, or ask them directly. The default
+ * preference is `"ask"`, but users can choose `"open"` or `"copy"` in their
+ * settings to avoid the prompt.
+ *
+ * @async
+ * @returns {Promise<UserOpenAction>} One of `"open"` or `"copy"`.
+ */
+async function askUserForOpenAction(): Promise<UserOpenAction> {
+  // first check if the user has set a default action
+  const prefAction =
+    vscode.workspace
+      .getConfiguration("shiny.shinylive")
+      .get<string>("openAction") || "ask";
+
+  if (["open", "copy"].includes(prefAction)) {
+    return prefAction as UserOpenAction;
+  }
+
+  const action = await vscode.window.showQuickPick(["open", "copy"], {
+    title: "Open or copy the ShinyLive link?",
+  });
+  return (action || "open") as UserOpenAction;
+}
+
+/**
+ * Consult the user's preferred Shinylive mode, or ask them directly. The
+ * default preference is `"ask"`, but users can choose `"app"` or `"editor"` in
+ * their settings to avoid the prompt.
+ *
+ * @async
+ * @returns {Promise<ShinyliveMode>} One of `"app"` or `"editor"`.
+ */
+async function askUserForAppMode(): Promise<ShinyliveMode> {
+  // first check if the user has set a default mode
+  const prefMode =
+    vscode.workspace
+      .getConfiguration("shiny.shinylive")
+      .get<string>("appMode") || "ask";
+
+  if (["app", "editor"].includes(prefMode)) {
+    return prefMode as ShinyliveMode;
+  }
+
+  // ask the user if they want to run the app or edit the code
+  const mode = await vscode.window.showQuickPick(["app", "editor"], {
```

Comment:
When you demoed this it took me a couple of seconds to remember what "app" and "editor" meant. Would it be hard to add a brief description to these choices?